### ISSUE TITLE
Sinon internal error TypeError: this.args[pos].apply is not a function

### ISSUE THREAD

**Issue by Alexsey** [sinonjs/sinon#1694]
* Sinon version: **4.3.0**
* Environment : **NodeJS**
* Other libraries you are using: **proxyquire**, **mocha**

**What did you expect to happen?**
Error with a descriptive message should be thrown (probably I'm doing something wrong?)

**What actually happens**
Internal Sinon error is thrown
```
TypeError: this.args[pos].apply is not a function
    at Object.callArgOnWith (node_modules\sinon\lib\sinon\call.js:114:24)
    at Object.callArgWith (node_modules\sinon\lib\sinon\call.js:109:28)
    at Function.spyApi.(anonymous function) (node_modules\sinon\lib\sinon\spy.js:408:47)
    at Context.it (foo.test.js:13:12)
    at <anonymous>
```

**How to reproduce**
`foo.js`:
```js
const rp = require('request-promise')

module.exports = () => rp('http://www.google.com')
```
`foo.test.js`
```js
const sinon = require('sinon')
const sandbox = sinon.createSandbox()
const proxyquire = require('proxyquire')

const rpStub = sandbox.stub()

const foo = proxyquire('./foo', {
    'request-promise': rpStub
})

it('should throw from Sinon internals', async () => {
    await foo()
    rpStub.callArg(0)
})
```

**Comment by fatso83** [comment]
This is a usage question. We try to keep this issue tracker for bugs. Try posting to StackOverflow where you'll find plenty of help. 

Your problem is twofold:
- you await the stub, but you forget to return a promise from the stub. See the docs for stub#resolves().
- you try to call the argument passed to the stub, which makes no sense as it is supposed to receive a string

**Comment by Alexsey** [comment]
@fatso83 my issue is not a question. It's about Sinon crash. Crash is something that should not happen even if something is doing wrong. If something is doing wrong then error with descriptive message should be thrown

Actually, it's a bug and to fix it we need to add checks in places like [this](https://github.com/sinonjs/sinon/blob/master/lib/sinon/call.js#L101), [this](https://github.com/sinonjs/sinon/blob/master/lib/sinon/call.js#L105) and [this](https://github.com/sinonjs/sinon/blob/master/lib/sinon/call.js#L114) that argument is a function. I can make a PR for that if you would reopen this issue

BTW `await` of non-promise is 100% valid

**Comment by fatso83** [comment]
Fair enough, I misunderstood the issue :-) Better, more informative error messages is of course an improvement we would very much welcome.

**Comment by mroderick** [comment]
> I can make a PR for that if you would reopen this issue

That would be very welcome 👍 

(love your github avatar)

**Comment by Alexsey** [comment]
@mroderick thank you :)

Here comes a PR https://github.com/sinonjs/sinon/pull/1699

**Comment by mroderick** [comment]
This has been published to npm as `sinon@4.4.2`

**Comment by mroderick** [comment]
Fixed with #1699
